Add new Puppeteer front-end e2e test: single product page

juliaamosova opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

1. Add a new front-end e2e test: single product page. 
2. Introduce 2 components:

- Create a simple product
- Create a variable product 

### Thoughts on components:

Why introduce the components? 

In order to run front-end e2e tests such as add item to the cart, we need to have a product created on the site. This can be done in different ways:

- **DB import** - this is how it currently works in WooCommerce Core e2e test suite but proved to be difficult to maintain over a period of time. I would like us not to go that route;

- **Create a new product in the test** - this is the approach I took in this PR. The downside is that we'd need to add it to each test that requires working with a product which can add longer tests run time over time. Note that this is how Gutenberg approached their tests - they are [creating a new post in each test](https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/specs/block-switcher.test.js#L14). 

- **Product import** - this is something we can explore and what could be a good solution. It should require less maintenance over time. The shared product import file can be created to be used for running tests both locally and on Travis. Import can be done in the setup test site stage which will result in products being available to interact with before tests start running. 

What are your thoughts on that? 

### How to test the changes in this Pull Request:

1. Setup e2e test environment as per README [in this PR](https://github.com/woocommerce/woocommerce/pull/24658/files) (you'd need to scroll down to README section). Note the changes around creating a new Customer user on the test site;
2. Enable Development mode by passing `--dev` flag to the `test:e2e` script in `./package.json` file as follows: `"test:e2e": "./tests/bin/e2e-test-integration.js --dev",`
3. Run e2e tests by using `npm run test:e2e` command;
4. Test headless by disabling `Development mode` and running `npm run test:e2e` command.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

juliaamosova committed:
Add new Puppeteer front-end e2e test: single product page

juliaamosova committed:
Introduce store owner logout flow

vedanshujain reviewed (commented):
Thanks for working on this @juliaamosova. I really liked the separation of CustomerFlow and StoreOwner flow and it helps in understanding what's going on just by looking at the code. 

Since this is my first time reviewing one of the e2e test PR, I have some minor overall suggestion as well:
1. I think we should rebase all of these branches with master, there are quite a few merge conflicts already and it will only increase.
2. Probably add another test command for dev like `"test:e2e-dev": "./tests/bin/e2e-test-integration.js --dev"` , that way we won't have to edit package.json everytime we want to test.

I added some minor comment and a blockage that I hit. Let me know what you think!

juliaamosova committed:
Resolve conflicts with migration branch

beaulebens commented on tests/e2e-tests/utils/flows.js:
```diff
@@ -7,7 +7,95 @@ const baseUrl = process.env.WP_BASE_URL;
 const WP_ADMIN_WC_SETTINGS = baseUrl + '/wp-admin/admin.php?page=wc-settings&tab=';
 
+const SHOP_PRODUCT = baseUrl + '/product/';
+const SHOP_CART_PAGE = baseUrl + '/cart/';
```
These are technically configurable in WC, right? Do we care? Or can we always just ensure that our test environment uses these same slugs?

juliaamosova replied:
We will need to make sure that our test environment has the same slugs setup. Otherwise, we would need to check for all kinds of variations of URLs.

beaulebens commented on tests/e2e-tests/specs/front-end/front-end-single-product.test.js:
```diff
@@ -0,0 +1,60 @@
+		await activatePlugin( 'woocommerce' );
+		await createSimpleProduct();
+		await createVariableProduct();
+		await StoreOwnerFlow.logout();
```
It's probably good practice just to ensure that the rest of the test runs "as a normal customer" vs anything special happening because there is an active admin session/logged in user.

juliaamosova replied:
IMO, yes, we do need to log out in order to run the rest of the test as a `customer` who is not logged in as an `admin` to match the typical shopping experience on WooCommerce site when someone buys a product.

beaulebens commented on tests/e2e-tests/utils/components.js:
```diff
@@ -0,0 +1,130 @@
+/**
+ * Create simple product.
+ */
+const createSimpleProduct = async () => {
```
Thinking ahead to other extensions etc re-using these core components, I imagine we'll want to be able to define the properties of a product when it's being created (title, price). And maybe return some kind of reference (like post-id) after creation?

juliaamosova replied:
Good call, thanks! I am not returning post-id for both simple and variable products once they are created.

juliaamosova committed:
Resolve conflicts with feature branch

juliaamosova committed:
Introduce postID after creating simple and variable products

juliaamosova commented on tests/e2e-tests/utils/components.js:
```diff
@@ -0,0 +1,130 @@
+	await StoreOwnerFlow.openNewProduct();
+
+	// Make sure we're on the add order page
+	await expect(page.title()).resolves.toMatch('Add new product');
```
Fixed it, thanks!

juliaamosova commented on tests/e2e-tests/utils/components.js:
```diff
@@ -0,0 +1,130 @@
+	await expect(page.title()).resolves.toMatch('Add new product');
+
+	// Set product data
+	await expect(page).toFill('#title', 'Simple product');
```
I agree with you @vedanshujain, thank you! I am now returning post-id for each product after it was created and then navigating to the URL with the correct post-id accordingly.

juliaamosova commented:
@vedanshujain thank you for your thoughtful review! 

>I think we should rebase all of these branches with master, there are quite a few merge conflicts already and it will only increase.

I did a rebase recently and will make sure to keep it up to date during the time left to complete the E2E test suite migration project.  

>Probably add another test command for dev like "test:e2e-dev": "./tests/bin/e2e-test-integration.js --dev" , that way we won't have to edit package.json everytime we want to test.

Good call, thank you! Added `test:e2e-dev` script as you suggested. Tests in the development mode can now be run using `test:e2e-dev`. 

>I added some minor comment and a blockage that I hit. Let me know what you think!

Thanks! I addressed the comments you left. This PR is now ready to be reviewed again.

juliaamosova commented on tests/e2e-tests/utils/components.js:
```diff
@@ -0,0 +1,130 @@
+
+	// Set some variation data
+	await uiUnblocked();
+	await uiUnblocked();
```
>In my local setup, I am not able to get past this point. If this is working fine on your end then I will look into more detail as to why this is failing.

We found out earlier that the test was not passing due to issue with variations loading slower than the test was running and not due to these calls. 

>Also, can you add a comment on why do we need to call this method twice?

The 1st one is being called [here](https://github.com/woocommerce/woocommerce/blob/d0491072e8a7d0a4535db0ec2bae4a6d43ff64fd/assets/js/admin/meta-boxes-product-variation.js#L641) when the alert letting us know that the variations were created out of attributes pops up. 

The 2nd one is [here](https://github.com/woocommerce/woocommerce/blob/d0491072e8a7d0a4535db0ec2bae4a6d43ff64fd/assets/js/admin/meta-boxes-product-variation.js#L402) after we dismiss the alert and wait for variations to be loaded.

vedanshujain reviewed (commented):
@juliaamosova Good work on this. Code looks good and both tests for Simple and Variable products are passing for me, but some different tests for Tax settings are failing with timeout errors. See below:
```
● WooCommerce Tax Settings › can set rate settings

    TimeoutError: Element input[name^="tax_rate_country[new-0"] not found

      waiting for function failed: timeout 500ms exceeded

  ● WooCommerce Tax Settings › can remove tax classes

    TimeoutError: Navigation timeout of 30000 ms exceeded

      125 | 		}
      126 |
    > 127 | 		await page.goto( settingsUrl, {
          | 		           ^
      128 | 			waitUntil: 'networkidle0',
      129 | 		} );
      130 | 	},
```

These looks unrelated to this PR, are they being fixed somewhere else, or do you suspect this is because of something on my local environment.

juliaamosova commented:
>Code looks good and both tests for Simple and Variable products are passing for me, but some different tests for Tax settings are failing with timeout errors. See below: 

>These looks unrelated to this PR, are they being fixed somewhere else, or do you suspect this is because of something on my local environment.

@vedanshujain it is possible that it is related to your local env but it is hard to tell. I am not getting these errors on my end. 

Tax settings test is now a part of https://github.com/woocommerce/woocommerce/pull/24218 - I think we should merge this PR and fix tax settings test in https://github.com/woocommerce/woocommerce/pull/24218 if we confirm that these errors are present after we set up Travis with Docker. What do you think?

vedanshujain reviewed (approved):
> Tax settings test is now a part of #24218 - I think we should merge this PR and fix tax settings test in #24218 if we confirm that these errors are present after we set up Travis with Docker. What do you think?

Makes sense. Lets go ahead and merge.
